tensorflow/core/kernels/fixed_length_record_reader_op.cc

```diff
@@ -28,45 +31,92 @@ namespace tensorflow {
+        encoding_(encoding) { }
 
   // On success:
-  // * input_buffer_ != nullptr,
-  // * input_buffer_->Tell() == header_bytes_
+  // * buffered_input_stream_ != nullptr,
+  // * buffered_input_stream_->Tell() == header_bytes_
   // * file_pos_limit_ == file size - footer_bytes_
   Status OnWorkStartedLocked() override {
     record_number_ = 0;
-    uint64 file_size = 0;
-    TF_RETURN_IF_ERROR(env_->GetFileSize(current_work(), &file_size));
-    file_pos_limit_ = file_size - footer_bytes_;
-
-    TF_RETURN_IF_ERROR(env_->NewRandomAccessFile(current_work(), &file_));
-
-    input_buffer_.reset(new io::InputBuffer(file_.get(), kBufferSize));
-    TF_RETURN_IF_ERROR(input_buffer_->SkipNBytes(header_bytes_));
+    if (encoding_ == "ZLIB" || encoding_ == "GZIP") {
+      const io::ZlibCompressionOptions zlib_options =
+          encoding_ == "ZLIB" ?
+          io::ZlibCompressionOptions::DEFAULT() :
+          io::ZlibCompressionOptions::GZIP();
+
+      file_pos_limit_ = -1;
+      footer_cache_.clear();
+
+      TF_RETURN_IF_ERROR(env_->NewRandomAccessFile(current_work(), &file_));
+      file_stream_.reset(new io::RandomAccessInputStream(file_.get()));
+      buffered_input_stream_.reset(new io::ZlibInputStream(file_stream_.get(),
+                                                          (size_t)kBufferSize,
+                                                          (size_t)kBufferSize,
+                                                          zlib_options));
+      TF_RETURN_IF_ERROR(buffered_input_stream_->SkipNBytes(header_bytes_));
+
+      TF_RETURN_IF_ERROR(buffered_input_stream_->ReadNBytes(footer_bytes_, &footer_cache_));
+    } else {
+      uint64 file_size = 0;
+      TF_RETURN_IF_ERROR(env_->GetFileSize(current_work(), &file_size));
+      file_pos_limit_ = file_size - footer_bytes_;
+
+      TF_RETURN_IF_ERROR(env_->NewRandomAccessFile(current_work(), &file_));
+      buffered_input_stream_.reset(new io::BufferedInputStream(file_.get(), kBufferSize));
+      TF_RETURN_IF_ERROR(buffered_input_stream_->SkipNBytes(header_bytes_));
+    }
     return Status::OK();
   }
 
   Status OnWorkFinishedLocked() override {
-    input_buffer_.reset(nullptr);
+    buffered_input_stream_.reset(nullptr);
     return Status::OK();
   }
 
   Status ReadLocked(string* key, string* value, bool* produced,
                     bool* at_end) override {
-    if (input_buffer_->Tell() >= file_pos_limit_) {
-      *at_end = true;
-      return Status::OK();
+    if (file_pos_limit_ < 0) {
+      // In case of encoded file (non-seekable), attempt to read
+      // record, and append record to cached footer, then pop head
+      // of cached footer as the record
+      Status s = buffered_input_stream_->ReadNBytes(record_bytes_, value);
+      if (!s.ok() && !errors::IsOutOfRange(s)) {
+        return s;
+      }
+
+      // EOF reached
+      if (value->size() == 0) {
+        file_pos_limit_ = buffered_input_stream_->Tell() - footer_bytes_;
```

Comment:
Thanks @saxenasaurabh. The updated PR includes additional comments for the explanation.